### PR TITLE
Refactor html-proofer

### PR DESCRIPTION
**Pull request by gjtorikian**
This refactors html-proofer into much more manageable and smaller chunks. Notably, it:

* Moves the logging logic into its own class.
* Moves the external URL checking logic into its own class.
* Renames the horrible `HTML::Proofer::Checks::Check` to a more sane `HTML::Proofer::Runner`.
* Simplifies custom test logic

It also namespaces Typheous/Hydra option passing; hence this will be a major version bump when merged.

Fixes https://github.com/gjtorikian/html-proofer/issues/131, fixes https://github.com/gjtorikian/html-proofer/issues/114, fixes https://github.com/gjtorikian/html-proofer/issues/129.

Although none of you are under any obligation whatsoever to do so, a review from @penibelst / @benbalter / @parkr might be nice, to make sure I'm not missing something.

### CODE REVIEW

File: bin/htmlproof [hunk 1]
```diff
@@ -29,31 +29,31 @@ Mercenary.program(:htmlproof) do |p|
...
+    options[:href_ignore] = opts['href_ignore'] unless opts['href_ignore'].nil?
+    options[:alt_ignore] = opts['alt_ignore'] unless opts['alt_ignore'].nil?
+    options[:file_ignore] = opts['file_ignore'] unless opts['file_ignore'].nil?
+    options[:disable_external] = opts['disable_external'] unless opts['disable_external'].nil?
+    options[:only_4xx] = opts['only_4xx'] unless opts['only_4xx'].nil?
+    options[:favicon] = opts['favicon'] unless opts['favicon'].nil?
+    options[:verbose] = opts['verbose'] unless opts['verbose'].nil?
+    options[:directory_index_file] = opts['directory_index_file'] unless opts['directory_index_file'].nil?
+    options[:validate_html] = opts['validate_html'] unless opts['validate_html'].nil?
+    options[:check_external_hash] = opts['check_external_hash'] unless opts['check_external_hash'].nil?
```
Comment (parkr) [inline bin/htmlproof]: can this be DRY'd?

**Comment by parkr** [comment]
:heart: :heart: :heart: :heart: :heart: :heart: :heart: :heart:

File: bin/htmlproof (hunk 1 above)
Comment (gjtorikian) [inline bin/htmlproof]: ugh probably. I was thinking about removing the bin entirely but I guess that's Not The Right Thing To Do :tm:.

**Comment by penibelst** [comment]
Really nice.

With smaller chunks we could sometimes test classes instead of the whole thing. I’ve added tests for the new Utils class in 700613208843c72b6d06b6400fb8616186040183.

**Comment by penibelst** [comment]
Nice run.

**Pull request merged**